log/log.go

```diff
@@ -2,121 +2,145 @@ package log
 
 import (
 	"fmt"
-	"log"
 	"os"
-	"strings"
 
 	"github.com/peak/s5cmd/flags"
 )
 
-// stdoutCh is used to synchronize writes to standard output. Multi-line
+// output is a structure for std logs.
+type output struct {
+	std     *os.File
+	message string
+}
+
+// outCh is used to synchronize writes to standard output. Multi-line
 // logging is not possible if all workers print logs at the same time.
-var stdoutCh = make(chan string, 10000)
+var outputCh = make(chan output, 10000)
 
-var Logger = New()
+// Logger is the global logger.
+var Logger *logger
 
+// Init inits global logger.
+func Init() {
+	Logger = New()
+}
+
+// logLevel is the level of Logger.
 type logLevel int
 
 const (
 	levelDebug logLevel = iota
 	levelInfo
-	levelSuccess
 	levelWarning
 	levelError
 )
 
+// String returns the string representation of logLevel.
 func (l logLevel) String() string {
 	switch l {
-	case levelSuccess:
-		return "+"
+	case levelInfo:
+		return ""
 	case levelError:
 		return "ERROR"
 	case levelWarning:
 		return "WARNING"
-	case levelInfo:
-		return "#"
 	case levelDebug:
 		return "DEBUG"
 	default:
 		return "UNKNOWN"
 	}
 }
 
+// levelFromString returns logLevel for given string. It
+// return `levelInfo` as a default.
+func levelFromString(s string) logLevel {
+	switch s {
+	case "debug":
+		return levelDebug
+	case "info":
+		return levelInfo
+	case "warning":
+		return levelWarning
+	case "error":
+		return levelError
+	default:
+		return levelInfo
+	}
+}
+
+// Message is an interface to print structured logs.
+type Message interface {
+	fmt.Stringer
+	JSON() string
+}
+
+// logger is a structure for logging messages.
 type logger struct {
 	donech chan struct{}
-	impl   *log.Logger
+	level  logLevel
 }
 
+// New creates new logger.
 func New() *logger {
+	level := levelFromString(*flags.LogLevel)
 	logger := &logger{
 		donech: make(chan struct{}),
-		impl:   log.New(os.Stdout, "", 0),
+		level:  level,
 	}
-	go logger.stdout()
+	go logger.out()
 	return logger
 }
 
-func (l *logger) printf(level logLevel, format string, args ...interface{}) {
-	raw := fmt.Sprintf(format, args...)
-	var msg string
-
-	switch level {
-	case levelError:
-		msg = fmt.Sprintf("-ERR %v", raw)
-	case levelWarning:
-		msg = fmt.Sprintf("%v %v", level, raw)
-	case levelSuccess, levelInfo:
-		msg = fmt.Sprintf("                   %v %v", level, raw)
-
+// printf prints message according to the given level, message and std mode.
+func (l *logger) printf(level logLevel, message Message, std *os.File) {
+	if level < l.level {
+		return
 	}
-	stdoutCh <- msg
-}
 
-func (l *logger) Debug(format string, args ...interface{}) {
-	if *flags.Verbose {
-		l.printf(levelDebug, format, args...)
+	if *flags.JSON {
+		outputCh <- output{
+			message: message.JSON(),
+			std:     std,
+		}
+	} else {
+		outputCh <- output{
+			message: fmt.Sprintf("%v %v", level, message.String()),
```

Comment:
while we're here, could we remove the leading whitespace on Info() calls? I mean this:

` download s3://bucket/obj`

Reply:
ignore this, you already mentioned this on another comment.

Reply:
done. also, tests are updated to use equals() rt suffix() for the strict check.